pkg/provisioner/conditionwatcher.go

```diff
@@ -0,0 +1,194 @@
+		AddFunc:    c.onCreate,
+		UpdateFunc: c.onUpdate,
+		DeleteFunc: c.onDelete,
+	})
+	azVolumeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.onCreate,
+		UpdateFunc: c.onUpdate,
+		DeleteFunc: c.onDelete,
+	})
+	azDriverNodeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.onCreate,
+		UpdateFunc: c.onUpdate,
+		DeleteFunc: c.onDelete,
+	})
+
+	go informerFactory.Start(ctx.Done())
+
+	synced := cache.WaitForCacheSync(ctx.Done(), azVolumeAttachmentInformer.HasSynced, azVolumeInformer.HasSynced, azDriverNodeInformer.HasSynced)
+	if !synced {
+		klog.Fatalf("Unable to sync caches for azuredisk CRIs")
+		os.Exit(1)
+	}
+
+	return &c
+}
+
+func (c *conditionWatcher) newConditionWaiter(ctx context.Context, objType objectType, objName string, conditionFunc func(obj interface{}, expectDelete bool) (bool, error)) (*conditionWaiter, error) {
+	klog.V(5).Infof("Adding a condition function for %s (%s)", objType, objName)
+	entry := waitEntry{
+		conditionFunc: conditionFunc,
+		waitChan:      make(chan waitResult, 1),
+	}
+
+	_, exists := c.waitMap.LoadOrStore(getTypedName(objType, objName), &entry)
+	if exists {
+		err := status.Errorf(codes.Aborted, "another wait operation in process for %s (%s)", objType, objName)
+		klog.Error(err)
+		return nil, err
+	}
+
+	return &conditionWaiter{
+		objType: objType,
+		objName: objName,
+		entry:   &entry,
+		watcher: c,
+	}, nil
+}
+
+func (c *conditionWatcher) onCreate(obj interface{}) {
+	c.handleEvent(obj, create)
+}
+
+func (c *conditionWatcher) onUpdate(_, newObj interface{}) {
+	c.handleEvent(newObj, update)
+}
+
+func (c *conditionWatcher) onDelete(obj interface{}) {
+	c.handleEvent(obj, delete)
+}
+
+func (c *conditionWatcher) handleEvent(obj interface{}, eventType eventType) {
+	metaObj := obj.(metav1.Object)
+
+	var objType objectType
+	switch obj.(type) {
+	case *v1alpha1.AzVolume:
+		objType = azVolumeType
+	case *v1alpha1.AzVolumeAttachment:
+		objType = azVolumeAttachmentType
+	case *v1alpha1.AzDriverNode:
+		objType = azDriverNodeType
+	default:
+		// unknown object type
+		klog.Errorf("unsupported object type %v", reflect.TypeOf(obj))
+		return
+	}
+
+	v, ok := c.waitMap.Load(getTypedName(objType, metaObj.GetName()))
+	if !ok {
+		return
+	}
+	klog.V(5).Infof("found a wait entry for object (%s)", metaObj.GetName())
+	entry := v.(*waitEntry)
+	conditionFunc := entry.conditionFunc
+	waitChan := entry.waitChan
+
+	result := waitResult{}
+
+	ok, err := conditionFunc(obj, eventType == delete)
+	klog.V(5).Infof("condition result: succeeded: %v, error: %v", ok, err)
+	// if err found, send error through channel
+	if err != nil {
+		result.err = err
+	} else if !ok {
+		// if no error was found but condition not met, return
+		return
+	}
+
+	runtimeObj, ok := obj.(runtime.Object)
+	if !ok {
+		result.err = status.Errorf(codes.Aborted, "object does not implement runtime.Object interface.")
```

Comment:
`codes.Internal` since this really isn't expected, and `codes.Aborted` is generally used to indicate there is already an operation being performed on the object.